### PR TITLE
Ce 1400 updates checks

### PR DESCRIPTION
**Pull request by kamilkoterba**
https://wikia-inc.atlassian.net/browse/CE-1400

Adding checks to verify data that was sent to ExactTarget actually get there.
- added post update checks for user_properties and city_list
- added classes for retrieving wiki data from ExactTarget
- removed unused user update method and tests for that method

@Wikia/community-engineering

### CODE REVIEW

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveUserHelper.php [hunk 1]
```diff
@@ -65,6 +70,42 @@ public function retrieveUserDataById( $iUserId ) {
...
+				$sProperty = null;
+				$sValue = null;
+				foreach ( $aProperties as $value ) {
+					if ( $value->Name === 'up_property' ) {
+						$sProperty = $value->Value;
+					} elseif ( $value->Name === 'up_value' ) {
+						$sValue = $value->Value;
+					}
+				}
+				$oExactTargetUserPropertiesData[$sProperty] = $sValue;
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveUserHelper.php]: Spacey brackets :)

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveUserHelper.php [hunk 2]
```diff
@@ -65,6 +70,42 @@ public function retrieveUserDataById( $iUserId ) {
...
+		if ( isset( $oUserPropertiesResult->OverallStatus ) && $oUserPropertiesResult->OverallStatus !== 'OK' ) {
+			throw new \Exception( $oUserPropertiesResult->OverallStatus );
+		}
+
+		if ( !empty( $oUserPropertiesResult->Results ) ) {
+			foreach ( $oUserPropertiesResult->Results as $oResult ) {
+				$aProperties = $oResult->Properties->Property;
+				$sProperty = null;
+				$sValue = null;
+				foreach ( $aProperties as $value ) {
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveUserHelper.php]: I think that the name ```$value``` is a little bit misleading. What about ```$oProperty``` or sth?

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveWikiHelper.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+		if ( isset( $oWikiResult->OverallStatus ) && $oWikiResult->OverallStatus !== 'OK' ) {
+			throw new \Exception( $oWikiResult->OverallStatus );
+		}
+
+		if ( !empty( $oWikiResult->Results->Properties->Property ) ) {
+			$aProperties = $oWikiResult->Results->Properties->Property;
+			$aExactTargetWikiData = [];
+			foreach ( $aProperties as $value ) {
+				$aExactTargetWikiData[$value->Name] = $value->Value;
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveWikiHelper.php]: +1 to spacey brackets ```[]``` and the doubts about the name of ```$value``` :)

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUpdateWikiTask.php [hunk 4]
```diff
@@ -7,7 +7,7 @@ class ExactTargetUpdateWikiTask extends ExactTargetTask {
 	 * Updates city_list data with fallback to create in ExactTarget
 	 * Builds $aDataExtensions array and passes it to a DE API interface.
 	 * @param  int $iCityId  A wiki's ID
-	 * @return void
+	 * @return string 'OK' | 'Error'
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUpdateWikiTask.php]: Why not a boolean?

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php [hunk 5]
```diff
@@ -22,14 +22,54 @@ public function verifyUserData( $iUserId ) {
...
+	 * @param string $sIgnoredProperty Name of property that doesn't have to be equal
+	 * @return bool true if equal
+	 * @throws \Exception when results are not equals
+	 */
+	protected function compareResults( $aExactTargetData, $aWikiaData, $sCallerName, $sIgnoredProperty = '' ) {
+		// Compare results
+		$aDiffWikiaDB = array_diff( $aExactTargetData, $aWikiaData );
+
+		if ( !empty( $sIgnoredProperty ) && count( $aDiffWikiaDB ) == 1 && isset( $aDiffWikiaDB[$sIgnoredProperty] ) ) {
+			// Lets continue to the end of function as it's acceptable that $sIgnoredProperty is the only difference
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php]: What about unsetting the ```$aDiffWikiaDB[ $sIgnoredProperty ]``` and then checking the length of the array?

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php [hunk 6]
```diff
@@ -22,14 +22,54 @@ public function verifyUserData( $iUserId ) {
...
 		} elseif ( count( $aDiffWikiaDB ) > 0 ) {
-			// Prepare diff and throw exception
-			$aDiffExactTarget = array_diff( $oWikiaUserData, $oExactTargetUserData );
+			// There are unacceptable differences. Prepare diff and throw exception
+			$aDiffExactTarget = array_diff( $aWikiaData, $aExactTargetData );
 			$sDiffRes = [];
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php]: Why indicating a ```string``` type?

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiDataVerificationTask.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+	 */
+	public function verifyWikiData( $iWikiId ) {
+		// Fetch data from ExactTarget
+		$oRetrieveWikiHelperTask = $this->getRetrieveWikiHelper();
+		$oExactTargetWikiData = $oRetrieveWikiHelperTask->retrieveWikiDataById( $iWikiId );
+		$this->info( __METHOD__ . ' ExactTarget wiki data record: ' . json_encode( $oExactTargetWikiData ) );
+
+		// Fetch data from Wikia DB
+		$oWikiHelper = $this->getWikiHelper();
+		$oWikiaWikiData = $oWikiHelper->getWikiDataArray( $iWikiId );
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiDataVerificationTask.php]: An ```Array``` type should be indicated.

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiTaskHelper.php [hunk 8]
```diff
@@ -66,7 +92,19 @@ public function prepareWikiDataExtensionForUpdate( $iCityId ) {
...
+				'Properties' => $aWikiData,
+		];
+
+		return $aWikiDataExtension;
+	}
+
+	public function getWikiDataArray( $iCityId ) {
 		/* Get wikidata from master */
 		$oWiki = \WikiFactory::getWikiById( $iCityId, true );
 		$aWikiData = [
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiTaskHelper.php]: Did you check [get_object_vars](http://php.net/manual/en/function.get-object-vars.php)?

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiDataVerificationTask.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+	/**
+	 * Compare results
+	 * @param array $aExactTargetData Array with results from ExactTarget
+	 * @param array$aWikiaData Array with results from Wikia DB
+	 * @param string $sCallerName name of function that called this one, needed for logs
+	 * @param string $sIgnoredProperty Name of property that doesn't have to be equal
+	 * @return bool true if equal
+	 * @throws \Exception when results are not equals
+	 */
+	protected function compareResults( $aExactTargetData, $aWikiaData, $sCallerName, $sIgnoredProperty = '' ) {
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiDataVerificationTask.php]: Could you merge this method with the one in ```ExactTargetUserDataVerificationTask.php```?

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiDataVerificationTask.php [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+		// Compare results
+		$aDiffWikiaDB = array_diff_assoc( $aExactTargetData, $aWikiaData );
+		if ( !empty( $sIgnoredProperty ) && count( $aDiffWikiaDB ) == 1 && isset( $aDiffWikiaDB[$sIgnoredProperty] ) ) {
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiDataVerificationTask.php]: The same as in ```ExactTargetUserDataVerificationTask.php```.

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveWikiHelper.php [hunk 11]
```diff
@@ -0,0 +1,83 @@
+<?php
+namespace Wikia\ExactTarget;
+
+class ExactTargetRetrieveWikiHelper extends ExactTargetTask {
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveWikiHelper.php]: Why not ```ExactTargetRetrieveWikiTask```? IMO a Helper shouldn't extend a Task class.

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveWikiHelper.php (hunk 3 above)
Comment (kamilkoterba) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveWikiHelper.php]: Per coding conventions we're adding spaces around array contents and don't around arrays keys
https://github.com/Wikia/guidelines/blob/master/PHP/CodingConventions.md#array-syntax

Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveWikiHelper.php]: Very true, my bad. :santa:

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php (hunk 5 above)
Comment (kamilkoterba) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetUserDataVerificationTask.php]: Good idea, thanks

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiTaskHelper.php (hunk 8 above)
Comment (kamilkoterba) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetWikiTaskHelper.php]: I'm not fetching all records here so probably it's better to have explicit list.

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetDataComparisonHelper.php [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+			$aDiffExactTarget = array_diff_assoc( $aWikiaData, $aExactTargetData );
+			$aDiffRes = [];
+			$aDiffRes[] = "--- Expected (Wikia DB)";
+			$aDiffRes[] = "+++ Actual (ExactTarget)";
+			foreach ( $aDiffExactTarget as $key => $val ) {
+				$aDiffRes[] = "- '$key' => '{$aDiffExactTarget[$key]}'";
+				$aDiffRes[] = "+ '$key' => '{$aDiffWikiaDB[$key]}'";
+			}
+			$this->debug( $sCallerName . ' ' . json_encode( $aDiffRes ) );
+			throw new \Exception( $sCallerName . " Verification failed, Record in ExactTarget doesn't match record in Wikia database.");
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetDataComparisonHelper.php]: Single quotes and a space before the last parentheses per conventions. :)

File: extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveWikiTask.php [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+		return null;
+	}
+
+	/**
+	 * Change dates format of city_created and city_last_timestamp fields retrieved from ExactTarget
+	 * to Y-m-d H:i:s that is format consistent with Wikia DB
+	 * @param $aExactTargetWikiData
+	 */
+	protected function formatDates( &$aExactTargetWikiData ) {
+		$sWikiaDateFormat = "Y-m-d H:i:s";
```
Comment (adamkarminski) [inline extensions/wikia/ExactTargetUpdates/tasks/ExactTargetRetrieveWikiTask.php]: Single quotes :)

**Pull request merged**